pkg/proxy/iptables/proxier.go

```diff
@@ -1220,28 +1220,35 @@ func (proxier *Proxier) syncProxyRules() {
 		}
 
 		// Capture load-balancer ingress.
-		fwChain := svcInfo.firewallChainName
-		for _, ingress := range svcInfo.LoadBalancerIPStrings() {
-			if hasEndpoints {
-				// create service firewall chain
-				if chain, ok := existingNATChains[fwChain]; ok {
-					proxier.natChains.WriteBytes(chain)
-				} else {
-					proxier.natChains.Write(utiliptables.MakeChainLine(fwChain))
-				}
-				activeNATChains[fwChain] = true
-				// The service firewall rules are created based on ServiceSpec.loadBalancerSourceRanges field.
-				// This currently works for loadbalancers that preserves source ips.
-				// For loadbalancers which direct traffic to service NodePort, the firewall rules will not apply.
+		if len(svcInfo.LoadBalancerIPStrings()) > 0 && hasEndpoints {
+			// We could elide the firewall chain when loadBalancerSourceRanges
+			// is empty, but that adds code and makes the flow harder to reason
+			// about, for not much gain.
+			fwChain := svcInfo.firewallChainName
+
+			// create service firewall chain
+			if chain, ok := existingNATChains[fwChain]; ok {
+				proxier.natChains.WriteBytes(chain)
+			} else {
+				proxier.natChains.Write(utiliptables.MakeChainLine(fwChain))
+			}
+			activeNATChains[fwChain] = true
+
+			for _, lbip := range svcInfo.LoadBalancerIPStrings() {
+				// The service firewall rules are created based on the
+				// loadBalancerSourceRanges field.  This only works for
+				// VIP-like loadbalancers that preserve source IPs.  For
+				// loadbalancers which direct traffic to service NodePort, the
+				// firewall rules will not apply.
```

Comment:
In one of my not-currently-part-of-the-PR commits I had moved this comment to the top of the `if len(svcInfo.LoadBalancerIPStrings()) > 0` section (and reworded it a little), because the comment applies to this section as a whole, not just this loop.